src/number_theory/pell.lean

```diff
@@ -398,4 +427,252 @@ end solution₁
+  ∃ a : solution₁ d, is_fundamental a :=
+begin
+  obtain ⟨a, ha₁, ha₂⟩ := exists_pos_of_not_is_square h₀ hd,
+  -- convert to `x : ℕ` to be able to use `nat.find`
+  have P : ∃ x' : ℕ, 1 < x' ∧ ∃ y' : ℤ, 0 < y' ∧ (x' : ℤ) ^ 2 - d * y' ^ 2 = 1,
+  { have hax := a.prop,
+    lift a.x to ℕ using (by positivity) with ax,
+    norm_cast at ha₁,
+    exact ⟨ax, ha₁, a.y, ha₂, hax⟩, },
+  classical, -- to avoid having to show that the predicate is decidable
+  let x₁ := nat.find P,
+  obtain ⟨hx, y₁, hy₀, hy₁⟩ := nat.find_spec P,
+  refine ⟨mk x₁ y₁ hy₁, (by {rw x_mk, exact_mod_cast hx}), hy₀, λ b hb, _⟩,
+  rw x_mk,
+  have hb' := (int.to_nat_of_nonneg $ zero_le_one.trans hb.le).symm,
+  have hb'' := hb,
+  rw hb' at hb ⊢,
+  norm_cast at hb ⊢,
+  refine nat.find_min' P ⟨hb, |b.y|, abs_pos.mpr $ y_ne_zero_of_one_lt_x hb'', _⟩,
+  convert b.prop using 1,
+  rw [← hb', sq_abs],
+end
+
+/-- The `n`th power of a fundamental solution is trivial if and only if `n = 0`. -/
+lemma pow_eq_one_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : a ^ n = 1 ↔ n = 0 :=
+begin
+  refine ⟨λ H, _, λ H, by rw [H, zpow_zero]⟩,
+  have H' : (a ^ n).y = 0 := by rw [H, y_one],
+  rcases lt_trichotomy 0 n with hn | rfl | hn,
+  { exact absurd H' (y_zpow_pos h.x_pos h.2.1 hn).ne', },
+  { refl, },
+  { have := y_zpow_pos h.x_pos h.2.1 (by rwa [lt_neg, neg_zero] : 0 < -n),
+    rw [zpow_neg, y_inv, H', neg_zero] at this,
+    exact false.elim (lt_irrefl _ this), }
+end
+
+/-- The `n`th power of a fundamental solution has positive `y` if and only if `n` is positive. -/
+lemma pow_y_pos_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : 0 < (a ^ n).y ↔ 0 < n :=
+begin
+  refine ⟨λ H, _, y_zpow_pos h.x_pos h.2.1⟩,
+  rcases lt_trichotomy 0 n with hn | rfl | hn,
+  { exact hn, },
+  { rwa [zpow_zero, y_one] at H, },
+  { have := y_zpow_pos h.x_pos h.2.1 (by rwa [lt_neg, neg_zero] : 0 < -n),
+    rw [zpow_neg, y_inv, lt_neg, neg_zero] at this,
+    exact false.elim (lt_irrefl _ $ H.trans this), }
+end
+
+/-- The `n`th power of a fundamental solution has negative `y` if and only if `n` is negative. -/
+lemma pow_y_neg_iff {a : solution₁ d} (h : is_fundamental a) (n : ℤ) : (a ^ n).y < 0 ↔ n < 0 :=
+begin
+  rw [← neg_neg n, zpow_neg, y_inv, neg_lt, neg_zero, neg_lt, neg_zero],
+  exact h.pow_y_pos_iff (-n),
+end
+
+/-- A power of a fundamental solution is never equal to the negative of a power of this
+fundamental solution. -/
+lemma pow_ne_neg_fundamental_pow {a : solution₁ d} (h : is_fundamental a) {n n' : ℤ} :
+   a ^ n ≠ -a ^ n' :=
+begin
+  intro hf,
+  apply_fun solution₁.x at hf,
+  have H := x_zpow_pos h.x_pos n,
+  rw [hf, x_neg, lt_neg, neg_zero] at H,
+  exact lt_irrefl _ ((x_zpow_pos h.x_pos n').trans H),
+end
+
+/-- The `x`-coordinate of a fundamental solution is a lower bound for the `x`-coordinate
+of any positive solution. -/
+lemma le_x {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 1 < a.x) :
```

Comment:
```suggestion
lemma x_le_x {a₁ : solution₁ d} (h : is_fundamental a₁) {a : solution₁ d} (hax : 1 < a.x) :
```

Reply:
OK.

Reply:
I'm also changing a few other similar ones (`le_y` to `y_le_y` and `mul_inv_lt_x` to `x_mul_inv_lt_x`).

Reply:
...or should it be `mul_inv_x_lt_x`?